### PR TITLE
Change "user" folder default location to AppData/Roaming/ on Windows systems

### PR DESCRIPTION
**Pull request by freiro**
Hello, this is my first PR, I tried to address #2177 and move the default location of the "user" folder to AppData/Roaming/Citra Emulator/(*).
If there is a "user" folder next to citra executable, that folder will take priority (portable mode).

Configuration changes should be handled with care to avoid problems between different versions.

Also note that SHGetKnownFolderPath works only on Windows Vista and later.

Thanks to @linkmauve for the suggestions he gave me yesterday.

Let me know what you think.

(*) After a bit of discussion we decided for Roaming instead of Local.

### CODE REVIEW

File: src/common/file_util.cpp [hunk 1]
```diff
@@ -594,6 +597,21 @@ std::string& GetExeDirectory() {
     }
     return exe_path;
 }
+
+std::string& AppDataLocalDirectory() {
+    // Windows Vista or later only
+    static std::string local_path;
```
Comment (lioncash) [inline src/common/file_util.cpp]: Hi, please don't use static lifetime variables for things like this. Also, this function should ideally return the string by value (the fact GetExeDirectory does this is something that should be corrected, but isn't within the scope of this PR).

You should be able to just do:

```cpp
PWSTR path = nullptr;
SHGetKnownFolderPath(FOLDERID_LocalAppData, 0, nullptr, &path);

std::string local_path = Common::UTF16ToUTF8(path);
CoTaskMemFree(path);

return local_path;
```

**Comment by j-selby** [comment]
Would FOLDERID_RoamingAppData be more appropriate here? Citra save data would not be considered machine dependent, and Roaming AppData is generally considered best-practice for this kind of data:

> [User Name\AppData\Roaming] is used for User Profile specific data, while the AppData\Local folder structure is used for Machine Specific data
https://social.technet.microsoft.com/Forums/windows/en-US/b4fd7ac2-8ad9-4a06-8c28-3f7993aa0272/appdata-local-vs-roaming?forum=w7itprogeneral

As you mentioned, this would mean that configuration data ([qt+sdl2]config.ini) would exist between instances. Would this be preferable when a Debug Citra might need different settings then a Release Citra? Of course, as you said, the portable configuration option would suffice, but I would think that would have to be clearly documented.

File: src/common/file_util.cpp (hunk 1 above)
Comment (freiro) [inline src/common/file_util.cpp]: For some reason yesterday compilation was failing, fixed 👍

**Comment by freiro** [comment]
> Would FOLDERID_RoamingAppData be more appropriate here? Citra save data would not be considered machine dependent, and Roaming AppData is generally considered best-practice for this kind of data

@j-selby  True, I choose Local AppData since the bleeding edge version does the same. It would be useful to have more feedback on this.

> As you mentioned, this would mean that configuration data ([qt+sdl2]config.ini) would exist between instances. Would this be preferable when a Debug Citra might need different settings then a Release Citra? Of course, as you said, the portable configuration option would suffice, but I would think that would have to be clearly documented.
Documentation would be the next step :)
Regarding "Debug Citra" version, I think that it shouldn't matter too much since it's enough to create the "user" folder next to citra(-qt).exe to use it.
Actually this might allow to switch quickly between two different configurations (release and debug possibly).

**Comment by j-selby** [comment]
@freiro Bleeding Edge also contains binary/library files in its installation, which would, by my above definition, be suited for Local I would guess (though, its own 'user' directory could live in Roaming). I'm likely overthinking this though.

File: src/common/file_util.cpp [hunk 2]
```diff
@@ -594,6 +597,19 @@ std::string& GetExeDirectory() {
     }
     return exe_path;
 }
+
+std::string AppDataLocalDirectory() {
+    std::string local_path;
+    if (local_path.empty()) {
```
Comment (freiro) [inline src/common/file_util.cpp]: Yes it will, since I removed static, same modifications would apply to GetExeDirectory.
Fixed.

File: src/common/file_util.cpp [hunk 3]
```diff
@@ -594,6 +597,18 @@ std::string& GetExeDirectory() {
     }
     return exe_path;
 }
+
+std::string AppDataLocalDirectory() {
+    std::string local_path;
+    PWSTR pw_local_path = nullptr;
+    // Only supported by Windows Vista or later
+    SHGetKnownFolderPath(FOLDERID_LocalAppData, 0, nullptr, &pw_local_path);
+    local_path = Common::UTF16ToUTF8(pw_local_path);
```
Comment (lioncash) [inline src/common/file_util.cpp]: local_path's declaration can be joined with the assignment here.

`// Freeing memory` isn't necessary either, it's self-evident that is what's being done.

File: src/common/file_util.cpp (hunk 2 above)
Comment (lioncash) [inline src/common/file_util.cpp]: std::string constructs by default to an empty string; it wouldn't be undefined. However, it's correct that the if statement wasn't needed since the static specifier was dropped.

File: src/common/file_util.cpp (hunk 3 above)
Comment (freiro) [inline src/common/file_util.cpp]: ok, what about std::string& GetExeDirectory(), shall we leave it to a different PR?

Comment (lioncash) [inline src/common/file_util.cpp]: yes

File: src/common/file_util.cpp [hunk 4]
```diff
@@ -671,6 +683,11 @@ const std::string& GetUserPath(const unsigned int DirIDX, const std::string& new
     if (paths[D_USER_IDX].empty()) {
 #ifdef _WIN32
         paths[D_USER_IDX] = GetExeDirectory() + DIR_SEP USERDATA_DIR DIR_SEP;
+        if (!FileUtil::IsDirectory(paths[D_USER_IDX])) {
+            paths[D_USER_IDX] =
+                AppDataLocalDirectory() + DIR_SEP + EMU_DATA_DIR DIR_SEP USERDATA_DIR DIR_SEP;
```
Comment (linkmauve) [inline src/common/file_util.cpp]: You still have one + too much here, you can concatenate `DIR_SEP` and `EMU_DATA_DIR` just by separating them with a space.

Other than that, LGTM.

Comment (freiro) [inline src/common/file_util.cpp]: Missed it, thanks :)

**Comment by emmauss** [comment]
wouldn't be better to let the user folder be in Documents\Citra and not appdata, like dolphin.some guys are gonna mess up their systems if they want to make changes to the user folder

**Comment by jroweboy** [comment]
@emmauss https://blogs.msdn.microsoft.com/patricka/2010/03/18/where-should-i-store-my-data-and-configuration-files-if-i-target-multiple-os-versions/

In short, configuration files for a single user thats not synced with active directory should go in AppData/Local

**Comment by emmauss** [comment]
@jroweboy but that would only apply to the config folder, but what about the nand files and sdmc files.users would regularly move things to and from there, so an easily accessible location would be good.

**Comment by jroweboy** [comment]
@emmauss 

    regularly move things to and from there

I disagree that people will regularly access these locations. There is a one time setup with a home folder dump, and after that, they wouldn't need to access the folder again for most use cases. The only other time I can see people wanting to access the folder is for syncing files between their real 3ds and citra such as save files. I personally don't have a strong opinion about where it goes, but I prefer to have it where Windows says it should go.

**Comment by yuriks** [comment]
Really disagree with the rationales here... the user directory should go in Roaming. It contains things like save files, sd contents, etc. which are not disposable data. Local should be used for things like caches and application installs.

Also, I think people want to access the directory all the time. Changing hidden config settings, transfering save games, adding files to the SD for use with homebrew, etc. etc.

**Comment by yuriks** [comment]
I would also prefer that this kind of decision be moved to the frontend (we could then just use Qt for example to get this) because I think this makes more sense in the front end. (As well as the filesystem layer in general) but that would be a large undertaking, and this isn't too bad.

**Comment by mailwl** [comment]
summon @hrydgard to this thread, he know all about multiplatform

**Comment by hrydgard** [comment]
Few people will run this on setups where it matters whether you put it in Roaming or Local , but yuriks is right. As long as portable mode still works for those who prefer it, I don't see why it matters. Don't know what other input I'm supposed to provide, heh.

**Comment by freiro** [comment]
Since users are currently having problems about this [(here)](https://discuss.citra-emu.org/d/985-save-files-gone ), I just committed a patch that selects **AppData/Roaming** as default folder.

Please change the PR title when you can.

**Comment by yuriks** [comment]
LGTM

**Comment by jroweboy** [comment]
Has this been tested in mingw builds yet? I looked up and saw that they were in the headers, but it'd be nice to know that it was tested in mingw64 first

https://github.com/Alexpux/mingw-w64/blob/master/mingw-w64-headers/include/combaseapi.h#L363
https://github.com/Alexpux/mingw-w64/blob/master/mingw-w64-headers/include/shlobj.h#L755

Other than that LGTM

**Comment by freiro** [comment]
Currently works on mingw without problems.

**Comment by linkmauve** [comment]
LGTM.

**Comment by yuriks** [comment]
Can we rename "Citra Emulator" to just "Citra" while we're breaking the directory location?

**Comment by freiro** [comment]
ok, now the folder is C:\Users\username\AppData\Roaming\Citra\.

**Comment by yuriks** [comment]
LGTM

File: src/common/file_util.cpp [hunk 5]
```diff
@@ -26,6 +26,9 @@
 #define stat _stat64
 #define fstat _fstat64
 #define fileno _fileno
+// Windows version, at least Vista is required to obtain AppData Path
+#define WINVER 0x0600
+#define _WIN32_WINNT 0x0600
```
Comment (JayFoxRox) [inline src/common/file_util.cpp]: Shouldn't this happen before the first include or via command line?
(And verified where necessary?)
